R/knitr-engine.R

```diff
@@ -0,0 +1,110 @@
+#' A reticulate Engine for Knitr
+#' 
+#' This provides a `reticulate` engine for `knitr`, suitable for usage when
+#' attempting to render Python chunks. Using this engine allows for shared state
+#' between Python chunks in a document -- that is, variables defined by one
+#' Python chunk can be used by later Python chunks.
+#' 
+#' The engine can be activated by setting (for example)
+#' 
+#' ```
+#' knitr::knit_engines$set(python = reticulate::eng_python)
+#' ```
+#' 
+#' Typically, this will be set within a document's setup chunk, or by the
+#' environment requesting that Python chunks be processed by this engine.
+#' 
+#' @param options
+#'   Chunk options, as provided by `knitr` during chunk execution.
+#'   
+#' @export
+eng_python <- function(options) {
+  
+  # extract the code to be run -- we'll attempt to run the code line by line
+  # and detect changes so that we can interleave code and output (similar to
+  # what one sees when executing an R chunk in knitr). to wit, we'll do our
+  # best to emulate the return format of 'evaluate::evaluate()'
+  code <- options$code
+  if (!length(code))
+    return(list())
+  
+  # helper function for extracting code
+  extract <- function(code, start, end) {
+    pasted <- paste(code[start:end], collapse = "\n")
+    sub("^[\\r\\n\\t ]*\\n", "", pasted)
+  }
+  
+  # indices tracking the current lines of code to be executed
+  start <- end <- 1
+  n <- length(code)
+  
+  # record where pending console output should be collected from
+  pending <- 1
+  
+  # actual outputs to be returned to knitr
+  outputs <- list()
+  
+  while (TRUE) {
+    
+    # check for end situation
+    if (start > n || end > n)
+      break
+    
+    # attempt evaluate of current chunk of code
+    snippet <- extract(code, start, end)
+    captured <- py_capture_output(
+      result <- tryCatch(
+        py_run_string(snippet, convert = FALSE),
+        error = identity
+      )
+    )
+    
+    # detect syntax errors: assume that this means we need to pass an extra
```

Comment:
Note to self: this probably won't work when submitting e.g. function definitions; e.g.

```
def foo:
    print 1
    print 2
```

If we go this route of interleaving input and output, we'll need some way of computing the span of a Python statement.